### ISSUE TITLE
TypeError: Cannot read property 'startRecordingFilesWithChangedResolutions' of undefined

### ISSUE THREAD

**Issue by amcasey** [microsoft/TypeScript#27876]
From 3.1.3:

```
Error processing request. Cannot read property 'startRecordingFilesWithChangedResolutions' of undefined
TypeError: Cannot read property 'startRecordingFilesWithChangedResolutions' of undefined
    at InferredProject.Project.updateGraph (tsserver.js:118761:38)
    at updateProjectIfDirty (tsserver.js:119449:45)
    at updateGraph (tsserver.js:119734:34)
    at Array.forEach (<anonymous>)
    at ProjectService.ensureProjectStructuresUptoDate (tsserver.js:119738:39)
    at ProjectService.doEnsureDefaultProjectForFile (tsserver.js:119722:22)
    at ProjectService.ensureDefaultProjectForFile (tsserver.js:119719:75)
    at IOSession.Session.getFileAndProjectWorker (tsserver.js:122451:87)
    at IOSession.Session.getFileAndProject (tsserver.js:122436:29)
    at IOSession.Session.getCompletions (tsserver.js:122605:31)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (tsserver.js:121665:61)
    at tsserver.js:123086:88
    at IOSession.Session.executeWithRequestId (tsserver.js:123077:28)
    at IOSession.Session.executeCommand (tsserver.js:123086:33)
    at IOSession.Session.onMessage (tsserver.js:123106:35)
    at Interface.<anonymous> (tsserver.js:124369:27)
    at emitOne (events.js:116:13)
    at Interface.emit (events.js:211:7)
    at Interface._onLine (readline.js:282:10)
    at Interface._normalWrite (readline.js:424:12)
    at Socket.ondata (readline.js:141:10)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at Pipe.onread (net.js:594:20)
```

**Comment by amcasey** [comment]
Command was `completionInfo`.

**Comment by amcasey** [comment]
I'll look but, for future reference, it's much easier to search for asserts that have message strings.

**Comment by sheetalkamat** [comment]
Even though there were not stacks for remove project I realized that 94df5167b0ecec912a5d63af2ec3b2f4c5b032f1 fixed lot of issues with incorrect script info still being attached to the project and then not being removed correctly. I think that should suffice. but as @amcasey suggests will add more info to assert for future.

**Comment by sheetalkamat** [comment]
Closing this to be re-opened if there is more info

**Comment by amcasey** [comment]
24 hour hit counts from VS Code:

TSVersion | Command | Count
--- | --- | ---
3.3.3333 | completionInfo | 39
3.3.3333 | definitionAndBoundSpan | 46
3.3.3333 | documentHighlights | 157
3.3.3333 | getApplicableRefactors | 555
3.3.3333 | jsxClosingTag | 9
3.3.3333 | quickinfo | 305
3.3.3333 | signatureHelp | 12
3.3.3 | completionInfo | 35
3.3.3 | definitionAndBoundSpan | 68
3.3.3 | documentHighlights | 70
3.3.3 | getApplicableRefactors | 274
3.3.3 | quickinfo | 181
3.3.1 | completionInfo | 9663
3.3.1 | definitionAndBoundSpan | 2308
3.3.1 | documentHighlights | 7171
3.3.1 | getApplicableRefactors | 39214
3.3.1 | getCodeFixes | 93
3.3.1 | jsxClosingTag | 533
3.3.1 | organizeImports | 1
3.3.1 | projectInfo | 4
3.3.1 | quickinfo | 14391
3.3.1 | signatureHelp | 1320
3.3.0-dev | completionInfo | 91
3.3.0-dev | definitionAndBoundSpan | 7
3.3.0-dev | documentHighlights | 43
3.3.0-dev | getApplicableRefactors | 381
3.3.0-dev | quickinfo | 211
3.3.0-dev | signatureHelp | 23

**Comment by amcasey** [comment]
@sheetalkamat is there telemetry for server asserts?  I couldn't find any messages containing "Found script Info still attached to project" (from 62419bb1039054997df40ced00c0c332b18bf62c).

**Comment by sheetalkamat** [comment]
This exception can be thrown only when https://github.com/microsoft/TypeScript/pull/32490/files#diff-efa731ca7c6279501f903db87403d2beL1090 is being thrown and which generally means there are more scriptinfos than we already removed. In most cases it means there was some exception before this which resulted in the invalid state of the projects and script infos and that's the exception we really care. So this should be closed with no action needed. 
Although while investigating this found out that we were emitting JSON stringify of scriptinfos which has circular structure so PR available to fix that as  #32490
